cordova-serve/src/server.js

```diff
@@ -0,0 +1,137 @@
+        function do404() {
+            console.log('404 ' + request.url);
+            response.writeHead(404, {'Content-Type': 'text/plain'});
+            response.write('404 Not Found\n');
+            response.end();
+        }
+
+        function do302(where) {
+            console.log('302 ' + request.url);
+            response.setHeader('Location', where);
+            response.writeHead(302, {'Content-Type': 'text/plain'});
+            response.end();
+        }
+
+        function do304() {
+            console.log('304 ' + request.url);
+            response.writeHead(304, {'Content-Type': 'text/plain'});
+            response.end();
+        }
+
+        function isFileChanged(path) {
+            var mtime = fs.statSync(path).mtime,
+                itime = request.headers['if-modified-since'];
+            return !itime || new Date(mtime) > new Date(itime);
+        }
+
+        var urlPath = url.parse(request.url).pathname;
+
+        var filePath;
+        if (opts.urlPathProcessor) {
+            var result = opts.urlPathProcessor(urlPath, request, response, do302, do404);
+            if (!result) {
+                return;
+            }
+            filePath = result.filePath;
+        }
+        if (!filePath) {
+            if (!root) {
+                throw new Error('No server root directory HAS BEEN specified!');
+            }
+            filePath = path.join(root, urlPath);
+        }
+
+        fs.exists(filePath, function (exists) {
+            if (!exists) {
+                do404();
+                return;
+            }
+            if (fs.statSync(filePath).isDirectory()) {
+                var index = path.join(filePath, 'index.html');
+                try {
+                    if (fs.statSync(index)) {
+                        filePath = index;
+                    }
+                } catch (e) {
+                }
+            }
+            if (fs.statSync(filePath).isDirectory()) {
+                if (!/\/$/.test(urlPath)) {
+                    do302(request.url + '/');
+                    return;
+                }
+                console.log('200 ' + request.url);
+                response.writeHead(200, {'Content-Type': 'text/html'});
+                response.write('<html><head><title>Directory listing of ' + urlPath + '</title></head>');
+                response.write('<h3>Items in this directory</h3>');
+                response.write('<ul>');
+                fs.readdirSync(filePath).forEach(function (file) {
+                    if (file) {
+                        response.write('<li><a href="' + file + '">' + file + '</a></li>\n');
+                    }
+                });
+
+                response.write('</ul>');
+                response.end();
+            } else if (!isFileChanged(filePath)) {
+                do304();
+            } else {
+                (opts.streamHandler && opts.streamHandler(filePath, request, response)) || stream(filePath, request, response);
```

Comment:
Hah, fair point. Though I'm tempted to make it:

    (opts.streamHandler || stream)(filePath, request, response);

Jut for kicks :smile: